kythe/docs/rfc/3339.md

```diff
@@ -0,0 +1,389 @@
+    types of analyses.
+-   `<output-encoding>`: an identifier indicating how the output will be
+    encoded, must be one of `json` or `protobuf` (JSON: string).
+
+Where the description refers to a protocol buffer message, the message is
+encoded using the canonical
+[JSON mapping](https://developers.google.com/protocol-buffers/docs/proto3#json)
+for the named message type.
+
+#### analysis-type
+
+Analysis types are described by precisely the same grammar as a
+[kythe fact](https://kythe.io/docs/kythe-storage.html#_a_id_termfact_a_fact),
+repeated here:
+
+```
+analysis-type =  "/" | 1*path
+path          =  "/" word
+word          =  1*{LETTER|DIGIT|PUNCT}
+LETTER        =  // Unicode letter, category Lu, Ll, Lt, Lm, or Lo
+DIGIT         =  // Unicode digit, category Nd
+PUNCT         =  [-.@#$%&_+:()]
+```
+ 
+Analysis types whose leading segment is `/kythe` are reserved for use by
+Kythe. An example analysis type might be `/kythe/index/java`, indicating the
+Kythe indexer for Java.
+
+### Methods
+
+A driver defines the following JSON-RPC methods:
+
+-   `init(InitRequest) → InitReply`
+
+    The `init` method is a protocol handshake, and must be the first method
+    called by the analyzer upon startup. The request specifies the protocol
+    version understood by the analyzer. If the environment understands the
+    requested version, it echoes the version in its reply.
+
+    If any other method is called prior to a successful invocation of `init`, or
+    if the environment doesn't recognize the protocol version requested, the
+    environment will report ProtocolError and close the input channel.
+
+    This document defines the protocol `"kythe1"`.
+
+    Example:
+
+        » {"jsonrpc":"2.0","id":1,"method":"init","params":{"protocol":"kythe1", "output-encoding":"json"}}
+        « {"jsonrpc":"2.0","id":1,"result":{"protocol":"kythe1"}}
+
+-   `analyze(AnalyzeRequest) → AnalyzeReply`
+
+    Request a fresh compilation record from the environment, blocking until one
+    is available. Once an analysis has started, it is designated as **pending**
+    until the analyzer sends a `done` notification (see below). The driver MUST
+    report ProtocolError and close the input channel if the analyzer issues an
+    `analyze` request while another analysis is pending.
+    
+    The reply includes the input file(s) on which to begin analysis, as well as
+    the output file to which to emit analysis output. It reflects some of the
+    kythe.proto.CompilationUnit structure. The driver (and driver environment)
+    MUST make contents available for each file referenced in the compilation
+    record at the file system path specified in the compilation record.
+
+    The output file should contain the content appropriate to the
+    analysis-type, encoded as specified in the `InitRequest`. For Kythe index
+    analyses, this would be `kythe.common.Entry` objects. Other analyzers might
+    emit a "findings" structure.
```

Comment:
This freedom is going to add complexity. Maybe the canonical output should be a gRPC Any proto? Or a finding proto with an Any field that can be used to encode Entries, etc.?

Reply:
Where do you see the complexity? I'd expect the driver would know a priori what type of data to expect.

Reply:
Maybe if we can get more people piling on https://github.com/protocolbuffers/protobuf/issues/5226 we can get a better resolution there (I'm not optimistic, but still...)